src/io/trailer.cc

```diff
@@ -0,0 +1,150 @@
+        return knowhere::Status::invalid_trailer;
+    }
+
+    if (trailer_ptr->GetCheckSum() != checksum) {
+        LOG_KNOWHERE_ERROR_ << "Checksum check fail.";
+        return knowhere::Status::invalid_trailer;
+    }
+    return knowhere::Status::success;
+}
+}  // namespace
+
+namespace knowhere {
+Status
+AddTrailerForMemoryIO(MemoryIOWriter& writer, const std::string& name, const Version& version) {
+    auto trailer_ptr = std::make_unique<Trailer>();
+    auto size = writer.tellg();
+    bool set_success = true;
+    set_success &= trailer_ptr->SetCheckSum(CalculateCheckSum(writer.data(), size));
+    set_success &= trailer_ptr->SetVersion(version.VersionNumber());
+    set_success &= trailer_ptr->SetIndexName(name);
+    if (!set_success) {
+        return Status::invalid_trailer;
+    }
+    writer.write(trailer_ptr->GetBytes(), KNOWHERE_TRAILER_SIZE);
+    return Status::success;
+}
+
+Status
+CheckTrailerForMemoryIO(MemoryIOReader& reader, const std::string& name) {
+    uint64_t bin_size = TRAILER_OFFSET(reader.size());
+    if (bin_size < 0) {
+        LOG_KNOWHERE_WARNING_ << "Trailer not exist in Binary.";
+        return Status::success;
+    }
+
+    auto trailer_ptr = std::make_unique<Trailer>();
+    auto pre_rp = reader.tellg();
+    reader.seekg(bin_size);
+    reader.read(trailer_ptr.get(), KNOWHERE_TRAILER_SIZE);
+    reader.seekg(pre_rp);
+
+    if (!trailer_ptr->TrailerValidCheck()) {
+        LOG_KNOWHERE_WARNING_ << "Trailer not exist in Binary.";
+        return Status::success;
+    }
+
+    auto checksum = CalculateCheckSum(reader.data(), bin_size);
+    return TrailerCheck(trailer_ptr, name, checksum);
+}
+
+Status
+AddTrailerForFiles(const std::vector<std::string>& files, const std::string& trailer_file, const std::string& name,
+                   const Version& version) {
+    auto trailer_ptr = std::make_unique<Trailer>();
+    bool set_success = true;
+    set_success &= trailer_ptr->SetCheckSum(GetFilesCheckSum(files));
+    set_success &= trailer_ptr->SetVersion(version.VersionNumber());
+    set_success &= trailer_ptr->SetIndexName(name);
+    if (!set_success) {
+        return Status::invalid_trailer;
+    }
+
+    std::ofstream writer(trailer_file.c_str(), std::ios::binary);
+    writer.write((char*)trailer_ptr->GetBytes(), KNOWHERE_TRAILER_SIZE);
+    writer.close();
+    return Status::success;
+}
+
+Status
+CheckTrailerForFiles(const std::vector<std::string>& files, const std::string& trailer_file, const std::string& name) {
+    std::ifstream reader(trailer_file.c_str(), std::ios::binary);
+    if (!reader) {
+        LOG_KNOWHERE_WARNING_ << "Trailer file not exist.";
+        return Status::success;
+    }
+    reader.seekg(0, std::ios::end);
+    auto fsize = reader.tellg();
+
+    reader.seekg(0, std::ios::beg);
+    if (fsize != KNOWHERE_TRAILER_SIZE) {
+        LOG_KNOWHERE_ERROR_ << "Trailer size (" << fsize << ")not correct.";
```

Comment:
should "trailer file exists but has incorrect size" indicate error instead of success?

Reply:
chatted offline, we should consider add a constant `max_version_without_trailer` to the knowhere versioning system, all indexes with a greater version should include trailer, or else error.